diff-hl-flydiff.el

```diff
@@ -0,0 +1,195 @@
+        (get-buffer-create " *diff-hl-diff*")))))
+
+
+(defun diff-hl-flydiff/update (old-fun &optional auto)
+  (unless (and auto
+            (or
+              (= diff-hl-flydiff-modified-tick (buffer-modified-tick))
+              (file-remote-p default-directory)
+              (not (buffer-modified-p))))
+    (funcall old-fun)))
+
+(defun diff-hl-flydiff/changes (&rest args)
+  (let* ((file buffer-file-name)
+          (backend (vc-backend file)))
+    (when backend
+      (let ((state (vc-state file backend)))
+        (cond
+          ((or
+             (buffer-modified-p)
+             (eq state 'edited)
+             (and (eq state 'up-to-date)
+               ;; VC state is stale in after-revert-hook.
+               (or revert-buffer-in-progress-p
+                 ;; Diffing against an older revision.
+                 diff-hl-reference-revision)))
+            (let (diff-auto-refine-mode res)
+              (with-current-buffer (diff-hl-flydiff-buffer-with-head)
+                (goto-char (point-min))
+                (unless (eobp)
+                  (ignore-errors
+                    (diff-beginning-of-hunk t))
+                  (while (looking-at diff-hunk-header-re-unified)
+                    (let ((line (string-to-number (match-string 3)))
+                           (len (let ((m (match-string 4)))
+                                  (if m (string-to-number m) 1)))
+                           (beg (point)))
+                      (diff-end-of-hunk)
+                      (let* ((inserts (diff-count-matches "^\\+" beg (point)))
+                              (deletes (diff-count-matches "^-" beg (point)))
+                              (type (cond ((zerop deletes) 'insert)
+                                      ((zerop inserts) 'delete)
+                                      (t 'change))))
+                        (when (eq type 'delete)
+                          (setq len 1)
+                          (cl-incf line))
+                        (push (list line len type) res))))))
+              (setq diff-hl-flydiff-modified-tick (buffer-modified-tick))
+              (nreverse res)))
+          ((eq state 'added)
+            `((1 ,(line-number-at-pos (point-max)) insert)))
+          ((eq state 'removed)
+            `((1 ,(line-number-at-pos (point-max)) delete))))))))
+
+;;;###autoload
+(define-minor-mode diff-hl-flydiff-mode
+  "Highlight diffs on-the-fly"
+  :lighter ""
+  :global t
+  (if diff-hl-flydiff-mode
+    (progn
+      (advice-add 'diff-hl-update :around #'diff-hl-flydiff/update)
+      (advice-add 'diff-hl-changes :override #'diff-hl-flydiff/changes)
+      (advice-add 'diff-hl-overlay-modified :override #'ignored)
+
+      (setq diff-hl-flydiff-timer
+        (run-with-idle-timer 0.3 t #'diff-hl-update t)))
+
+    (advice-remove 'diff-hl-update #'diff-hl-flydiff/update)
+    (advice-remove 'diff-hl-changes #'diff-hl-flydiff/changes)
+    (advice-remove 'diff-hl-overlay-modified #'ignored)
```

Comment:
Fixed, but I can't actually change the commit message for a few hours. Anyway, I think this is still blocked by the Emacs bug.